src/hb-ot-layout.cc

```diff
@@ -1703,6 +1703,63 @@ hb_ot_layout_get_size_params (hb_face_t       *face,
 
   return false;
 }
+
+struct hb_position_single_dispatch_t :
+       hb_dispatch_context_t<hb_position_single_dispatch_t, bool>
+{
+  static return_t default_return_value () { return false; }
+  bool stop_sublookup_iteration (return_t r) const { return r; }
+
+  private:
+  template <typename T, typename ...Ts> auto
+  _dispatch (const T &obj, hb_priority<1>, Ts&&... ds) HB_AUTO_RETURN
+  ( obj.position_single (std::forward<Ts> (ds)...) )
+  template <typename T, typename ...Ts> auto
+  _dispatch (const T &obj, hb_priority<0>, Ts&&... ds) HB_AUTO_RETURN
+  ( default_return_value () )
+  public:
+  template <typename T, typename ...Ts> auto
+  dispatch (const T &obj, Ts&&... ds) HB_AUTO_RETURN
+  ( _dispatch (obj, hb_prioritize, std::forward<Ts> (ds)...) )
+};
+
+hb_position_t
+hb_ot_layout_get_optical_bound (hb_font_t      *font,
+				unsigned        lookup_index,
+				hb_direction_t  direction,
+				hb_codepoint_t  glyph)
+{
```

Comment:
My immediate questions, trying to use this api:
- How do I interpret the value that is returned ? Is it 'how far the glyph should extend into the margin' or something else?
- Where do I get this lookup_index you want?
- Can we add some boolean for 'has optical bound information' ? Or maybe that would be answered by trying to find that lookup_index?

Reply:
> * How do I interpret the value that is returned ? Is it 'how far the glyph should extend into the margin' or something else?

Right. Though, it's *negative* for extending into the margin. That's just how the numbers in the font were and I kept them that way. So basically that's how much you add to the x_offset and advance_width, for the left side, and you add to advance_width, for the right side.

> * Where do I get this lookup_index you want?

Ultimately from `hb_ot_layout_feature_get_lookups`. You need to get feature index first.

> * Can we add some boolean for 'has optical bound information' ? Or maybe that would be answered by trying to find that lookup_index?

Yeah, if there was no lookup_index you know...  Or even earlier, there would be no feature_index.

Reply:
I've got lucky with hb_ot_layout_collect_lookups

Reply:
We should probably turn this discussion into a doc comment

Reply:
Theoretical question: what would I do if ltbd returned more than one lookup ? try all until I find the glyph ?

Reply:
> Theoretical question: what would I do if ltbd returned more than one lookup ? try all until I find the glyph ?

Correct.

Reply:
> Theoretical question: what would I do if ltbd returned more than one lookup ? try all until I find the glyph ?

The loop you have ignores current script/lang. You might want to adjust it but is more work of course. You might as well keep it and loop over the lookups.